File: cocktail_bar.c
#include"cocktail_bar.h"

int main()
{	
	Cocktail* cocktail = NULL;
	Ingredient* stock = NULL;
	yaml(&cocktail, &stock, "load");
	int nb_cocktail = count_cocktail(cocktail);
	char check[25], who[25];	
	printf("Who are you:\n \t- bartender ?\n \t- customer(default) ?\n");
	scanf("%s", who);
	person = (strstr("bartender", who))? 1: 0;
	do
	{
		menu(cocktail, stock, nb_cocktail);
		printf("\nDo you want to make a cocktail (yes/no) ?\n");
		scanf("%s", check);
	}while (strstr("yes", check));
	(person != 1)? printf("It will make you %0.2f €\n", turnover): 0;
	printf("Have a nice day :)\n");
	
	cyaml_free(&config, &stock_sequence_schema, stock, 0);
	cyaml_free(&config, &cocktail_sequence_schema, cocktail, 0);
	return 0;
}


int count_cocktail(Cocktail* cocktail)
{
	int i = 0, nb_cocktail = 0;
	while(cocktail[i].name != NULL)
	{
		nb_cocktail++;
		i++;
	}
	return nb_cocktail;
}

void reset_yaml()
{
	Cocktail* init_cocktail;
	Ingredient* init_stock;
	cyaml_err_t  err_cocktail, err_stock;
	unsigned cocktail_count, stock_count;
	err_cocktail = cyaml_load_file("init_cocktail.yml", &config, &cocktail_sequence_schema,
			(void **)&init_cocktail, &cocktail_count);

	err_stock = cyaml_load_file("init_stock.yml", &config, &stock_sequence_schema,
			(void **)&init_stock, &stock_count);

	if (err_cocktail != CYAML_OK || err_stock != CYAML_OK)
		exit(EXIT_FAILURE);

	err_cocktail = cyaml_save_file("tmp_cocktail.yml", &config, &cocktail_sequence_schema,
			(void*)init_cocktail, __INIT_COCKTAIL);

	err_stock = cyaml_save_file("tmp_stock.yml", &config, &stock_sequence_schema,
			(void*)init_stock, STOCK);

	if (err_cocktail != CYAML_OK || err_stock != CYAML_OK)
		exit(EXIT_FAILURE);
}

void yaml(Cocktail** cocktail, Ingredient** stock, char* mode)
{
	cyaml_err_t err_cocktail, err_stock;
	unsigned cocktail_count, stock_count;
	err_cocktail = (!strcmp(mode, "load"))?cyaml_load_file("tmp_cocktail.yml", &config, &cocktail_sequence_schema,
			(void **)cocktail, &cocktail_count): cyaml_save_file("tmp_coktail.yml", &config, &cocktail_sequence_schema,
				(void*)*cocktail, count_cocktail(*cocktail));

	err_stock = (!strcmp(mode, "load"))?cyaml_load_file("tmp_stock.yml", &config, &stock_sequence_schema,
			(void **)stock, &stock_count): cyaml_save_file("tmp_stock.yml", &config, &stock_sequence_schema,
				(void*)*stock, STOCK);
	if (err_cocktail != CYAML_OK || err_stock != CYAML_OK)
		exit(EXIT_FAILURE);
}

void menu(Cocktail* cocktail, Ingredient* stock, int nb_cocktail)
{
	char choice[25]; 
	printf("\nDo you want to make:\n\t- an alcoholic cocktail,\n\t- non_alcoholic cocktail,\n\t- your own.\n");
	(person == 1)? printf("\t- modify the stock\n"): 0;
	scanf("%s", choice);
	//non_alcoholic sub_string of alcoholic
	if (strstr("alcoholic", choice))
		display_cocktail(cocktail, stock, true, nb_cocktail);
	else if(strstr("non_alcoholic", choice))
		display_cocktail(cocktail, stock, false, nb_cocktail);
	else if(strstr("own", choice))
		homemade(cocktail, stock);
	else if(strstr("modify", choice))
		stock_var(stock);
	else
	{
		printf("Sorry, you've entered a wrong choice.\n");
		menu(cocktail, stock, nb_cocktail);
	}
}	

//We display all the cocktails alcoholic or non-acloholic among the card according to the customer choice
void display_cocktail(Cocktail* cocktail, Ingredient* stock, bool value, int nb_cocktail)
{
	int amount[2];
	Specs* specs = malloc(nb_cocktail * sizeof(Specs));
	for (int i = 0; i < nb_cocktail; i++)
	{		
		//value refers to the absence (false) or presence (true) of alcohol in the cocktails
		specs[i] = specificity(cocktail[i].ingredient, cocktail[i].nb_ingredient);
		if (cocktail[i].alcoholic == value && cocktail[i].name != NULL && availability(cocktail[i], stock) == true)
			printf("\n- %s(%0.1f°, %0.1f g (sugar)), %0.2f cL, %0.2f €", cocktail[i].name,
					specs[i].alcoholic, specs[i].sugar, specs[i].volume, specs[i].price);
	}
	select_(cocktail, stock, amount, nb_cocktail);
	turnover += specs[amount[0]].price * amount[1];
	free(specs);
}

Specs specificity(Ingredient *ingredient, size_t nb_ingredient)
{
	float price = 0.0, volume = 0.0, sugar = 0.0, alcohol_volume = 0.0;
	for (size_t i = 0; i < nb_ingredient; i++)
	{
		price += ingredient[i].price * ingredient[i].quantity * 0.001;
		volume += ingredient[i].quantity;
		sugar += ingredient[i].sugar;
		if (ingredient[i].alcohol > 0)
			alcohol_volume += ingredient[i].quantity * ingredient[i].alcohol * 0.01;
	}
	Specs specs;
	specs.price = 1.1 * price;
	specs.volume = 0.1 * volume;
	specs.sugar = sugar;
	specs.alcoholic = alcohol_volume * 100 / volume;
	return specs;
}

//We get the cocktail chosen by the customer 
void select_(Cocktail* cocktail, Ingredient* stock, int* amount, int nb_cocktail)
{
	char selection[50], test[25];
	printf("\n\nWhich cocktail do you want ?\n\n\t");
	scanf("%s", selection);
	for (int i = 0; i < nb_cocktail; i++)
	{
		if (strstr(cocktail[i].name, selection))
		{
			if (availability(cocktail[i], stock) == false)
			{
				printf("\nSorry, I don't have enough ingredients to make a %s.\n", cocktail[i].name);
				select_(cocktail, stock, amount, nb_cocktail);
			}
			printf("\nA %s, very good choice it's my favorite cocktail.\n", cocktail[i].name);
			printf("How many %s do you want?\n", cocktail[i].name);
			scanf("%d", &amount[1]);
			amount[0] = i;
			if (person == 0)
				quantity_Less(cocktail[i].ingredient, stock, cocktail[i].nb_ingredient);
			else
			{
				printf("Do you want to see and/or modify the stock (yes/no)\n\n\t");
				scanf("%s", test);
				if (strstr("yes", test))
					stock_var(stock);
			}
			return;
		}
	}
	printf("\nExcuse me, I don't know this cocktail.\n");
	select_(cocktail, stock, amount, nb_cocktail);
}

//We see if the bartender can create the cocktail among the available ingredients in the stock
bool availability(Cocktail cocktail, Ingredient* stock)
{
	for (size_t i = 0; i < cocktail.nb_ingredient ; i++)
	{
		for (int j = 0; j < STOCK; j++)
		{
			if (cocktail.ingredient[i].name == stock[j].name && cocktail.ingredient[i].quantity > stock[j].quantity)
				return false;
		}
	}
	return true;
} 

void homemade(Cocktail* cocktail, Ingredient* stock)
{
	int i, count = 0;
	bool check = true;
	float quantity;
	char choice[25], test[25];
	Ingredient* p_ingredient = malloc(STOCK * sizeof(Ingredient));
	printf("List of Ingredients : \n");
	for(i = 0; i < STOCK; i++)
	{
		(stock[i].quantity != 0 )? printf("\n\t - %s : %0.2f € (kg/L)",stock[i].name,stock[i].price): 1;
	}
	do
	{
		while(check == true)
		{
			printf("\nWhat ingredient do you want to add?\n");
			scanf("%s", choice);
			for(i = 0; i < STOCK; i++)
			{
				if(!strcmp(stock[i].name, choice))
				{
					check = false;
					break;
				}
			}	
			if (check == false)
			{
				p_ingredient[count].price = stock[i].price;
				printf("What quantity of %s do you want ?\n", choice);
				scanf("%f", &quantity);
				if (quantity <= stock[i].quantity)
				{
					p_ingredient[count].name = malloc(strlen(choice) * sizeof(char));
					strcpy(p_ingredient[count].name, choice);
					p_ingredient[count].quantity = quantity;
					count++;
				}
				else
					printf("Sorry, we don't have enough %s\n", choice);
			}
			else
				printf("Sorry, I don't understand\n");
		}
		printf("Do you want another ingredient (yes/no) ?\n");
		scanf("%s", test);
	}while(strstr("yes", test));
	if (person == 0)
		quantity_Less(p_ingredient, stock, count);
	else
	{
		printf("\nDo you want to see and/or modify the stock (yes/no)\n\n\t");
		scanf("%s", test);
		if (strstr("yes", test))
			stock_var(stock);
	}
	Specs specs = specificity(p_ingredient, count);
	int amount;
	printf("\nHow many of your marvellous cocktail do you want ?\n");
	scanf("%d", &amount);
	turnover += specs.price * amount;
	printf("Do you want to add your cocktail to the list (yes/no)?\n");
	scanf("%s", test);
	if (strstr("yes", test))
		save_cocktail(cocktail, stock, p_ingredient, count);
}

void save_cocktail(Cocktail* cocktail, Ingredient* stock, Ingredient* p_ingredient, int size)
{
	int amongus = 0, nb_cocktail = count_cocktail(cocktail);
	if (realloc(cocktail, (nb_cocktail + 1) * sizeof(Cocktail)) == NULL)
		return;
	printf("Enter a name for your creation: ");
	scanf("%s", cocktail[nb_cocktail].name);
	printf("\nname: %s", cocktail[nb_cocktail].name);
	cocktail[nb_cocktail].ingredient = malloc(size * sizeof(Ingredient));
	for (int i = 0; i < size; i++)
	{
		cocktail[nb_cocktail].ingredient[i].name = p_ingredient[i].name;
		cocktail[nb_cocktail].ingredient[i].quantity = p_ingredient[i].quantity;	
		for (int j = 0; j < STOCK; j++)
		{
			if (cocktail[nb_cocktail].ingredient[i].name == stock[j].name)
			{
				cocktail[nb_cocktail].ingredient[i].alcohol = stock[j].alcohol;
				cocktail[nb_cocktail].ingredient[i].sugar = stock[j].sugar;
				cocktail[nb_cocktail].ingredient[i].price = stock[j].price * cocktail[nb_cocktail].ingredient[i].quantity * 0.001;
				if (cocktail[nb_cocktail].ingredient[i].alcohol > 0)
					amongus = 1;
				printf("\n\t- name: %s, \n\t- quantity: %f, \n\t- alcohol: %f, \n\t- sugar: %f, \n\tprice: %f", cocktail[nb_cocktail].ingredient[i].name,
						cocktail[nb_cocktail].ingredient[i].quantity, cocktail[nb_cocktail].ingredient[i].alcohol, cocktail[nb_cocktail].ingredient[i].sugar,
						cocktail[nb_cocktail].ingredient[i].price);
				break;
			}
		}
	}
	cocktail[nb_cocktail].alcoholic = (amongus == 1)? true: false;
	(cocktail[nb_cocktail].alcoholic == true)? printf("\nalcoholic: true"): printf("\nalcoholic: false");
	yaml(&cocktail, &stock, "save");
}

void quantity_Less(Ingredient* ingredient, Ingredient* stock, size_t nb_ingredient)
{
    for (size_t i = 0; i < nb_ingredient; i++)
    {
		for(int j = 0 ; j < STOCK ; j++)
        {
            if(strstr(stock[j].name, ingredient[i].name))
            {
				stock[j].quantity -= ingredient[i].quantity;
			}
		}
	}
}

void stock_var(Ingredient* stock)
{
	int i;
	float quantity;
	char test[25], ingredient[25], modify[25];
	printf("\n The stock contains:\n");
	for(i = 0; i < STOCK; i++)
	{
		printf("\t - %s (%0.3f kg/L): %0.2f €\n",stock[i].name, stock[i].quantity / 1000, stock[i].price);
	}
	do
	{
		printf("\nYou want to change the stock of which ingredient:\n\t");
		scanf("%s", ingredient);
		for (i = 0; i < STOCK; i++)
		{
			if (strstr(stock[i].name, ingredient))
			{
				printf("\nYou want to: \n\t- add some %s (default) ;\n\t- remove some %s;\n", stock[i].name, stock[i].name);
				scanf("%s", modify);
				do
				{
					printf("\nWhich quantity ?\n\t");
					scanf("%f", &quantity);
					if (quantity > stock[i].quantity && strstr("remove", modify))
						printf("There is just %0.3f Kg %s in the stock", stock[i].quantity * 0.001, stock[i].name);
				}while(quantity > stock[i].quantity && strstr("remove", modify));
				stock[i].quantity = (strstr("remove", modify))? stock[i].quantity - quantity: stock[i].quantity + quantity;
				quantity = 0.0;
			}
		}
		printf("\nDo you want to continue to modify the stock (yes/no)\n\t");
		scanf("%s", test);
	}while(strstr("yes", test));
	printf("\nNow the stock contains:\n");
	for(i = 0; i < STOCK; i++)
	{
		printf("\t - %s (%0.3f Kg/L): %0.2f €\n",stock[i].name, stock[i].quantity / 1000, stock[i].price);
	}
}

File: cocktail_bar.py
import yaml as yml

class Ingredient:
    def __init__(self, name, alcohol, quantity, sugar, price):
        self.name = name
        self.alcohol = alcohol
        self.quantity = quantity
        self.sugar = sugar
        self.price = price
    @classmethod
    def from_yaml(cls, y):
        return cls(
            y['name'],
            y['alcohol'],
            y['quantity'],
            y['sugar'],
            y['price']
        )
class Cocktail:
    def __init__(self, name, alcoholic, *ingredients):
        self.name = name
        self.alcoholic = alcoholic
        self.ingredients = ingredients
    @classmethod
    def from_yaml(cls, y):
        return cls(
            y['name'],
            y['alcoholic'],
            *[Ingredient.from_yaml(i) for i in y['ingredient']] 
        )

class Specs:
    def __init__(self):
        self.alcohol = 0
        self.volume = 0
        self.sugar = 0
        self.price = 0
class Order:
    def __init__(self):
        self.content = []
        self.amount = []


#global vars:

person, turnover = 0, 0


def yaml(cocktail, stock):
    with open(r'cocktail.yml') as yml_cocktail, open(r'stock.yml') as yml_stock:
        loaded_cocktail = yml.load(yml_cocktail, Loader = yml.FullLoader)
        loaded_stock = yml.load(yml_stock, Loader = yml.FullLoader)
        return [[Cocktail.from_yaml(i) for i in loaded_cocktail],
        [Ingredient.from_yaml(i) for i in loaded_stock]]


def main():
    global person
    who = input("Who are you:\n \t- bartender ?\n \t- customer(default) ?\n")
    if who in "bartender":
        person = 1
        if input("Do you want to reset the cocktail card and the stock (yes/no) ?\n") in "yes":
            reset_yaml()
    cocktail, stock = yaml([], [])
    order = Order()
    while True:
        menu(cocktail, stock, order)
        if input("\nDo you want to make a cocktail (yes/no) ?\n") not in "yes":
            break

    if len(order.content) > 0: 
        # order_var(order)
        if not person:
            print("It will make you {}$\n".format(turnover))
    print("Have a nice day :)\n")
    return 0

def menu(cocktail, stock, order):
    choice = input("Do you want to make:\n\t- an alcoholic cocktail,\n\t- a non alcoholic cocktail,\n\t- your own\n" +
    ("\t- modify the stock\n" if person == 1 else ""))
    if choice in "alcoholic":
        displayCocktail(cocktail, stock, order, True)
    elif choice in "non alcoholic":
        displayCocktail(cocktail, stock, order, False)
    elif choice in "own":
        homemade(cocktail, stock, order)
    elif choice in "modify" and person == 1:
        stock_var(stock)
    else:
        print("Sorry, you've entered a wrong choice.\n")
        menu(cocktail, stock, order)

def displayCocktail(cocktail, stock, order, value):
    specs = []
    for i in range(len(cocktail)):
        specs.append(specificity(cocktail[i].ingredients))
        if cocktail[i].alcoholic == value and cocktail[i].name != None and availability(cocktail[i].ingredients, stock) == True:
            print("\n- {0}({1:.1f}°, {2:.1f}° g (sugar)), {3:.2f} cL, {4:.2f} $".format(cocktail[i].name,
					specs[i].alcohol, specs[i].sugar, specs[i].volume, specs[i].price))
    select(cocktail, stock, order)


def specificity(ingredients):
    specs = Specs()
    for i in range(len(ingredients)):
        specs.price += ingredients[i].price * ingredients[i].quantity * 0.001
        specs.volume += ingredients[i].quantity
        specs.sugar += ingredients[i].sugar
        if ingredients[i].alcohol > 0:
            specs.alcohol += ingredients[i].quantity * ingredients[i].alcohol * 0.01
    specs.price *= 1.1
    specs.alcohol *= 100 / specs.volume
    return specs

def availability(ingredients, stock):
    for i in range(len(ingredients)):
        for j in range(len(stock)):
            if ingredients[i].name in stock[j].name and ingredients[i].quantity > stock[j].quantity:
                return False
    return True

def select(cocktail, stock, order):
    selection = input("\n\nWhich cocktail do you want ?\n\n\t")
    for i in range(len(cocktail)):
        if selection in cocktail[i].name:
            if availability(cocktail[i].ingredients, stock) == False:
                print("\nSorry, I don't have enough ingredients to make a {}\n".format(cocktail[i].name))
                select(cocktail, stock, order)
            print("\nA {}, very good choice it's my favourite cocktail.\n".format(cocktail[i].name))
            if order.content == None:
                return
            for j in range(len(order.content)):
                if order.content[j].name in cocktail[i].name:
                    order.amount[j] += 1 
                    break
            if cocktail[i].name in order.content:
                order.content.append(cocktail[i])
                order.amount.append(1)
            quantityLess(cocktail[i].ingredients, stock)
            if person == 1:
                if input("Do you want to see and//or modify the stock (yes/no)") in "yes":
                    stock_var(stock)
            return
    print("\nExcuse me, I don't know this cocktail.\n") 
    select(cocktail, stock, order)

def quantityLess(ingredients, stock):
    for i in range(len(ingredients)):
        for j in range(len(stock)):
            if ingredients[i].name in stock[j].name:
                stock[j].quantity -= ingredients[i].quantity
            

main()